include/envoy/filesystem/filesystem.h

```diff
@@ -25,37 +30,55 @@ class File {
    *
    * @return bool whether the open succeeded
    */
-  virtual Api::SysCallBoolResult open() PURE;
+  Api::IoCallBoolResult open() {
+    if (isOpen()) {
+      return resultSuccess<bool>(true);
+    }
+
+    openFile();
+    return -1 != fd_ ? resultSuccess<bool>(true) : resultFailure<bool>(false, errno);
+  }
```

Comment:
It looks like this and some of the others below were moved from the impl file. Typically we don't have much implementation in the interface and keep it abstract. What was the rationale behind this move?

Reply:
There are three options that I can see here:
1. Have duplicate code for the Windows and Linux implementations.
2. Change this interface to be abstract, add another class that derives from the abstract class with the common implementation and use this new class as the base class for the Windows and Linux implementations.
3. Put the common code in the base class (as implemented).

I prefer to avoid duplicate code and this is why option 1 is a bad idea.
I don't like option 2 either since it's less readable and it can also influence a bit on the performance (another layer in the virtual table).
This is why option 3 was chosen.

I thought about splitting filesystem.h to header and implementation files but I saw that there are no `.cc` files under the include directory.

Reply:
Typical Envoy style would be closest to #2, I would say. I definitely agree that duplicating the implementation is not desired.

In Envoy, the stuff under `include/` is generally 100% abstract interface classes. There's no `.cc` files in there. There are a few non-abstract bits under `include` and they are all there for performance reasons. Do we have concrete evidence that performance is an important factor here?

Reply:
I don't have specific evidence that the performance is important here. Just thought that we would like to make the performance as good as possible where it's quite easy to do so. 
I'll implement option 2.

Reply:
Done.